services/venice-server/src/main/java/com/linkedin/venice/listener/ReadQuotaEnforcementHandler.java

```diff
@@ -330,27 +307,77 @@ private void updateQuota(PartitionAssignment partitionAssignment) {
       LOGGER.warn(
           "Routing data changed on quota enforcement handler with 0 replicas assigned to this node, removing quota for resource: {}",
           topic);
-      storeVersionBuckets.remove(topic);
+      storeVersionRateLimiters.remove(topic);
       return;
     }
     long quotaInRcu = storeRepository.getStore(Version.parseStoreFromKafkaTopicName(partitionAssignment.getTopic()))
         .getReadQuotaInCU();
-    storeVersionBuckets.compute(topic, (k, v) -> {
-      long newRefillAmount = calculateRefillAmount(quotaInRcu, thisNodeQuotaResponsibility);
-      if (v == null || v.getAmortizedRefillPerSecond() * enforcementIntervalSeconds != newRefillAmount) {
-        // only replace the existing bucket if the difference is greater than 1
-        return tokenBucketfromRcuPerSecond(quotaInRcu, thisNodeQuotaResponsibility);
-      } else {
-        return v;
-      }
-    });
-    String storeName = Version.parseStoreFromVersionTopic(topic);
-    stats.setStoreTokenBucket(storeName, getBucketForStore(storeName));
+    storeVersionRateLimiters.compute(
+        topic,
+        (k, v) -> getRateLimiter(
+            topic,
+            quotaInRcu,
+            thisNodeQuotaResponsibility,
+            v,
+            storeVersionRateLimiterType,
+            quotaEnforcementIntervalInMs,
+            enforcementCapacityMultiple,
+            clock));
   }
 
-  private long calculateRefillAmount(long totalRcuPerSecond, double thisBucketProportionOfTotalRcu) {
-    long totalRefillAmount = totalRcuPerSecond * enforcementIntervalSeconds;
-    return (long) Math.ceil(totalRefillAmount * thisBucketProportionOfTotalRcu);
+  /**
+   * Get the rate limiter for the store version. If the rate limiter is already created and the quota is the same,
+   * return the existing rate limiter.
+   */
+  static VeniceRateLimiter getRateLimiter(
+      String storeVersionName,
+      long quotaInRcu,
+      double thisNodeQuotaResponsibility,
+      VeniceRateLimiter currentRateLimiter,
+      RateLimiterType rateLimiterType,
+      int quotaEnforcementIntervalInMs,
+      int enforcementCapacityMultiple,
+      Clock clock) {
+    // If the rate limiter is already created and the quota is the same, return the existing rate limiter
+    long newQuota = (long) Math.ceil(quotaInRcu * thisNodeQuotaResponsibility);
+    if (currentRateLimiter != null && currentRateLimiter.getGlobalPermitsCount() == quotaInRcu
+        && currentRateLimiter.getThisNodeResponsibility() == thisNodeQuotaResponsibility) {
+      return currentRateLimiter;
+    }
+
+    VeniceRateLimiter newRateLimiter;
+    if (rateLimiterType == RateLimiterType.EVENT_THROTTLER_WITH_SILENT_REJECTION) {
+      LOGGER.info(
```

Comment:
Let's not spam our logs with these `Creating x rate limiter for store version x with quota x`. If we want to know the latest rate limiter quota for a store version we can emit that as a metric. The `rateLimiterType` is a server config and won't change from store to store so if we want we can just emit it as a single log message when ReadQuotaEnforcementHandler is initialized.

Reply:
Good point